File: backend/Minimarket.Api/Services/UserService.cs
using Minimarket.Api.DTOs;
using Minimarket.Api.Helpers;
using Minimarket.Api.Mapping;
using Minimarket.Api.Models;
using Minimarket.Api.Repositories;

namespace Minimarket.Api.Services;

public class UserService(IUserRepository userRepository, IPasswordHasher passwordHasher) : IUserService
{
    public async Task<IReadOnlyCollection<UserDto>> GetAllAsync() =>
        (await userRepository.GetAllAsync()).Select(x => x.ToDto()).ToList();

    public async Task<(bool Success, string? Error, UserDto? User)> CreateAsync(CreateUserDto dto)
    {
        if (await userRepository.GetByUsernameAsync(dto.Username.Trim()) is not null)
        {
            return (false, "El nombre de usuario ya esta registrado.", null);
        }

        var user = new User
        {
            FullName = dto.FullName.Trim(),
            Username = dto.Username.Trim(),
            PasswordHash = passwordHasher.Hash(dto.Password),
            Role = dto.Role.Trim().ToLowerInvariant(),
            IsActive = dto.IsActive
        };

        await userRepository.AddAsync(user);
        await userRepository.SaveChangesAsync();
        return (true, null, user.ToDto());
    }

    public async Task<LoginResponseDto?> LoginAsync(LoginRequestDto dto)
    {
        var user = await userRepository.GetByUsernameAsync(dto.Username.Trim());
        if (user is null || !user.IsActive || !passwordHasher.Verify(dto.Password, user.PasswordHash))
        {
            return null;
        }

        return new LoginResponseDto(user.Id, user.FullName, user.Username, user.Role);
    }
}

File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/service/UserService.java
package com.minimarket.api.service;

import com.minimarket.api.dto.*;
import com.minimarket.api.entity.User;
import com.minimarket.api.repository.UserRepository;
import com.minimarket.api.util.DtoMapper;
import com.minimarket.api.util.PasswordHasher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordHasher passwordHasher;

    public UserService(UserRepository userRepository, PasswordHasher passwordHasher) {
        this.userRepository = userRepository;
        this.passwordHasher = passwordHasher;
    }

    public List<UserDto> getAll() {
        return userRepository.findAllByOrderByFullNameAsc()
            .stream()
            .map(DtoMapper::toDto)
            .toList();
    }

    @Transactional
    public ServiceResult<UserDto> create(CreateUserDto dto) {
        var username = dto.username().trim();
        if (userRepository.findByUsername(username).isPresent()) {
            return ServiceResult.failure("El nombre de usuario ya esta registrado.");
        }

        var user = new User();
        user.setFullName(dto.fullName().trim());
        user.setUsername(username);
        user.setPasswordHash(passwordHasher.hash(dto.password()));
        user.setRole(dto.role().trim().toLowerCase());
        user.setIsActive(dto.isActive());

        var saved = userRepository.save(user);
        return ServiceResult.success(DtoMapper.toDto(saved));
    }

    public LoginResponseDto login(LoginRequestDto dto) {
        var user = userRepository.findByUsername(dto.username().trim()).orElse(null);
        if (user == null || !Boolean.TRUE.equals(user.getIsActive()) || !passwordHasher.verify(dto.password(), user.getPasswordHash())) {
            return null;
        }

        return new LoginResponseDto(user.getId(), user.getFullName(), user.getUsername(), user.getRole());
    }
}
